GVRf/Framework/framework/src/main/jni/objects/lightlist.cpp

```diff
@@ -0,0 +1,372 @@
+/* Copyright 2015 Samsung Electronics Co., LTD
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/***************************************************************************
+ * Holds scene objects. Can be used by engines.
+ ***************************************************************************/
+
+#include "engine/renderer/renderer.h"
+#include "objects/lightlist.h"
+#include "objects/scene.h"
+#include "shaders/shader.h"
+
+#define LIGHT_ADDED 1
+#define LIGHT_REMOVED 2
+#define SHADOW_CHANGED 4
+#define REBUILD_SHADERS 8
+
+namespace gvr {
+
+LightList::~LightList()
+{
+    if (mLightBlock)
+    {
+        delete mLightBlock;
+        mLightBlock = nullptr;
+    }
+#ifdef DEBUG_LIGHT
+    LOGD("LIGHT: deleting light block");
+#endif
+}
+
+int LightList::getLights(std::vector<Light*>& lightList) const
+{
+    std::lock_guard < std::recursive_mutex > lock(mLock);
+    for (auto itc = mClassMap.begin(); itc != mClassMap.end(); ++itc)
+    {
+        const std::vector<Light*>& lights = itc->second;
+        for (auto itl = lights.begin(); itl != lights.end(); ++itl)
+        {
+            Light* light = *itl;
+            lightList.push_back(light);
+        }
+    }
+    return lightList.size();
+}
+
+/*
+ * Adds a new light to the scene.
+ * Return true if light was added, false if already there or too many lights.
+ */
+bool LightList::addLight(Light* light)
+{
+    std::lock_guard < std::recursive_mutex > lock(mLock);
+
+    auto it2 = mClassMap.find(light->getLightClass());
+    if (it2 != mClassMap.end())
+    {
+        std::vector<Light*>& lights = it2->second;
+        light->setLightIndex(lights.size());
+        lights.push_back(light);
+    }
+    else
+    {
+        std::pair<std::string, std::vector<Light*>> pair;
+
+        light->setLightIndex(0);
+        pair.first = light->getLightClass();
+        pair.second.push_back(light);
+        mClassMap.insert(pair);
+    }
+    mDirty |= LIGHT_ADDED | REBUILD_SHADERS;
+#ifdef DEBUG_LIGHT
+    LOGD("LIGHT: %s added to scene", light->getLightClass());
+#endif
+    return true;
+}
+
+/*
+ * Removes an existing light from the scene.
+ * Return true if light was removed, false if light was not in the scene.
+ */
+bool LightList::removeLight(Light* light)
+{
+    std::lock_guard < std::recursive_mutex > lock(mLock);
+
```

Comment:
why do we need recursive lock here?

Reply:
Because we have a situation where it hangs if I don't use a recursive lock. It has something to do with clear being called in an inconvenient place between tests. If you take the recursive lock out and run the lighting and shadow tests, you will find the hang (unless something else fixed it).